HPCC-15544 FirstN/Enth lookahead issues in loop

jakesmith opened the pull request:
Global FirstN and Enth's lookahead was not being correctly
restarted.
Avoid reconstruction if helper based settings haven't changed
and ensure lookahead object is restartable.

Signed-off-by: Jake Smith jake.smith@lexisnexis.com

jakesmith commented:
@ghalliday - please review

HPCCSmoketest commented:
Automated Smoketest 
Sha: 364b077c94d0d026cb193b2da74532826831529e
Build: success
Install hpccsystems-platform-community_6.0.0-rc4.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

ghalliday commented on thorlcr/activities/enth/thenthslave.cpp:
```diff
@@ -94,20 +99,32 @@ class BaseEnthActivity : public CSlaveActivity, implements ILookAheadStopNotify
         denominator = validRC(helper->getProportionDenominator());
         numerator = validRC(helper->getProportionNumerator());
     }
+    virtual void stop() override
```
@jakesmith 
I can see why you are doing it, but this code looks less clean than the previous code.  It seems to be heading in the wrong direction with the stranding code.

Could you possibly set inputStream to originalInputStream in start(), and have similar code to the new code to cache lookupHead, and assign it to inputStream?

jakesmith committed:
HPCC-15544 FirstN/Enth lookahead issues in loop

Global FirstN and Enth's lookahead was not being correctly
restarted.
Avoid reconstruction if helper based settings haven't changed
and ensure lookahead object is restartable.

Signed-off-by: Jake Smith <jake.smith@lexisnexis.com>

jakesmith commented:
@ghalliday - I've simplified the approach - please re-review

HPCCSmoketest commented:
Automated Smoketest 
Sha: 2ec5eac3d2be8368605d313fa9dc9dfb48bc564f
Build: success
Install hpccsystems-platform-community_6.0.0-rc4.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

ghalliday commented on thorlcr/activities/enth/thenthslave.cpp:
```diff
@@ -82,8 +83,8 @@ class BaseEnthActivity : public CSlaveActivity, implements ILookAheadStopNotify
-            IStartableEngineRowStream *lookAhead = createRowStreamLookAhead(this, originalInputStream, queryRowInterfaces(input), ENTH_SMART_BUFFER_SIZE, true, false, RCUNBOUND, this, &container.queryJob().queryIDiskUsage());
-            setLookAhead(0, lookAhead); // NB: this is post base start()
+            lookAhead = createRowStreamLookAhead(this, originalInputStream, queryRowInterfaces(input), ENTH_SMART_BUFFER_SIZE, true, false, RCUNBOUND, this, &container.queryJob().queryIDiskUsage());
+            replaceInputStream(0, lookAhead); // NB: this is post base start()
```
I think this is going to leak.

jakesmith replied:
You're right, previous method used to capture the lookahead in an owned, now links it only.
Will fix.

ghalliday commented on thorlcr/activities/enth/thenthslave.cpp:
```diff
@@ -103,11 +104,23 @@ class BaseEnthActivity : public CSlaveActivity, implements ILookAheadStopNotify
+        PARENT::stop();
+
+        // restore original inputStream if lookAhead was installed, to avoid base start spuriously starting previously installed lookahead
+        if (lookAhead)
```
trivial: This could be unconditional.  I'm not sure if it is clearer conditional or unconditional.

ghalliday commented on thorlcr/activities/firstn/thfirstnslave.cpp:
```diff
@@ -251,9 +251,14 @@ class CFirstNSlaveGlobal : public CFirstNSlaveBase, implements ILookAheadStopNot
         IStartableEngineRowStream *lookAhead = createRowStreamLookAhead(this, originalInputStream, queryRowInterfaces(input), FIRSTN_SMART_BUFFER_SIZE, isSmartBufferSpillNeeded(this), false,
                                                                               maxRead, this, &container.queryJob().queryIDiskUsage()); // if a very large limit don't bother truncating
-        setLookAhead(0, lookAhead);
+        replaceInputStream(0, lookAhead);
```
will also leak

ghalliday commented:
@jakesmith that looks cleaner.  I think it may currently leak, please can you check.

jakesmith commented:
@ghalliday - fixed leak and simplified a bit more - please check.

ghalliday commented on thorlcr/activities/firstn/thfirstnslave.cpp:
```diff
@@ -249,11 +244,17 @@ class CFirstNSlaveGlobal : public CFirstNSlaveBase, implements ILookAheadStopNot
+    {
+        PARENT::stop();
+        if (originalInputStream)
+            replaceInputStream(0, originalInputStream.getClear());
```
This leaks the lookahead value that is returned.

jakesmith replied:
Doh - yes

HPCCSmoketest commented:
Automated Smoketest 
Sha: 3f14e935bb5d0d9177f6a85c36ab8630e17dc6e2
Build: success
Install hpccsystems-platform-community_6.0.0-rc4.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

jakesmith committed:
HPCC-15544 Avoid leak in stop()

Signed-off-by: Jake Smith <jake.smith@lexisnexis.com>

jakesmith commented:
@ghalliday - back to you

HPCCSmoketest commented:
Automated Smoketest \n  a Test a \nSha: 6a3369ff554b28a385ec3579772e570b132e8f65\nBuild: success\nInstall hpccsystems-platform-community_6.0.0-rc4.el6.x86_64.rpm\nHPCC Start: OK\nHPCC Stop: OK\nHPCC Uninstall: OK\n
